Bundle with tsup and replace babel-jest

bryanjtc opened the pull request:
Bundles code and some dependencies with tsup.
Replaces babel-jest with @swc/core. With swc tests should run faster than using babel.
Adds require.resolve for jest-junit and jest-serializer-html.
Adds support for esm only projects that don't use a custom config file.
Tested locally using test-storybook:ci commands and some projects that I have with storybook and test-runner.
<!-- GITHUB_RELEASE PR BODY: canary-version -->
<details>
  <summary>📦 Published PR as canary version: <code>1.0.0--canary.224.3c57b66.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @storybook/test-runner@1.0.0--canary.224.3c57b66.0
  # or 
  yarn add @storybook/test-runner@1.0.0--canary.224.3c57b66.0
  ```
</details>
<!-- GITHUB_RELEASE PR BODY: canary-version -->

bryanjtc commented:
@yannbf Can you check this?

bryanjtc commented:
@yannbf What do you think of these changes?

yannbf commented:
Hey @bryanjtc thanks for your contribution! Sorry, thing have been busy with SB beta. We'll take a look!

ndelangen commented on package.json:
```diff
@@ -16,35 +16,27 @@
+  "types": "dist/index.d.ts",
   "bin": {
-    "test-storybook": "./bin/test-storybook.js"
+    "test-storybook": "./dist/test-storybook"
```
Is it OK to have no file-extension here?

bryanjtc replied:
Yes. Node automatically uses test-storybook.js for cjs projects and test-storybook.mjs for esm projects.

ndelangen commented on src/playwright/transformPlaywrightJson.ts:
```diff
@@ -86,7 +86,7 @@ export const transformPlaywrightJson = (index: Record<string, any>) => {
       const storyTests = stories.map((story) => makeDescribe(story.name, [makeTest(story)]));
       const program = t.program([makeDescribe(stories[0].title, storyTests)]);
-
+      // @ts-ignore
```
I think this should have a reason added:

```suggestion
      // @ts-ignore (why is this OK? who should fix this / when?)
```

ndelangen replied:
Don't actually commit ☝️, should have the actual reason (which I do not know)

bryanjtc replied:
I got rid of almost all ts-ignore comments. and fixed the errors.

yannbf replied:
Probably `ts-expect-error` instead

ndelangen commented on tsconfig.json:
```diff
@@ -13,7 +13,7 @@
     "rootDir": "./src",
     "skipLibCheck": true,
-    "target": "es5",
+    "target": "es6",
```
```suggestion
    "target": "es2020",
```

bryanjtc replied:
Done

ndelangen commented:
Just some finer questions/points really. Looks really good, in fact, @bryanjtc I'd love to get that `@swc/jest` change in the storybook monorepo some day too.

FYI: I have a chain or PRs on there removing babel: (all except for unit test 👎 ) (and code-mods 👍 )
https://github.com/storybookjs/storybook/pull/20156

ndelangen commented on tsup.config.ts:
```diff
@@ -0,0 +1,21 @@
+  },
+  {
+    clean: true,
+    entry: ['bin'],
```
We should just move the file into `src` then?

bryanjtc replied:
Done

ndelangen commented:
@yannbf Is there no CI set up for this project?

yannbf commented:
> @yannbf Is there no CI set up for this project?

There is, I guess it doesn't run on forks? I'll take a look at this tomorrow

bryanjtc commented:
> > @yannbf Is there no CI set up for this project?
> 
> There is, I guess it doesn't run on forks? I'll take a look at this tomorrow

The pr needs a label for the ci to run. Can you add it?

ndelangen commented:
I don't know what labels to apply to make the right things happen 😞

bryanjtc commented:
@yannbf Any update on getting ci to work?

ndelangen commented:
@yannbf is this supposed to work:
```
[SB] ERR! Module not found: Error: Can't resolve './stories/stress-test' in '/home/runner/work/test-runner/test-runner'
[SB] ERR! ModuleNotFoundError: Module not found: Error: Can't resolve './stories/stress-test' in '/home/runner/work/test-runner/test-runner'
```

bryanjtc commented:
I enabled some ci workflows for pull requests and modified them to run in node 16 and 18. http-server doesn't work with node 18, so I replaced it with serve. Snapshots were updated, nothing major, it just removed some empty lines.

yannbf commented:
Hey @bryanjtc thank you for all you work in this! I am currently on holidays, so I'll be checking this thoroughly once I'm back next week, I hope you don't mind!

yannbf commented:
Hey @bryanjtc I took a quick look, and to make a proper canary release I made another branch off of this one, but somehow that made this fork be able to release as well, but given this PR had "major" in it, it made a 1.0.0 version 🤦

With a canary available (and subsequent, as you make new pushes to this branch), let's test it out in our projects.

When testing, the binary did not exist anymore, so I got `Command "test-storybook" not found` when running `yarn test-storybook` in a project which was previously working.
I like what you did to the `test-storybook.js` file, but unfortunately it's not so easily understood when inspecting it via node_modules and easily modifiable so people could try quick fixes. But I guess that's expected if we want to support cjs/esm. Maybe if the prebundled code came from a vendors file, it would still be relatively simple to understand 🤔

bryanjtc commented:
Node should be able to find the bin file without extension. By default it should use `test-storybook.js` and for ESM only projects `test-storybook.mjs`. When specifying the extension, ESM projects expect `test-storybook.js` to be esm code. There are two solutions for this, changing the extension to `.cjs` or not specifying the extension. I can undo the prebundling for the dependencies of test-storybook.ts, so that it can be easier to read and modify in `node_modules`.

ndelangen commented:
@yannbf let's discuss this Monday?

bryanjtc commented:
Any update on this?

bryanjtc commented:
Any issues trying to re-create the pr? @yannbf

JReinhold commented:
This looks great!

@ndelangen will add you to the Storybook organization, and @yannbf will redo the PR to fix the CI issue.

ndelangen commented:
I'll re-open this PR

ndelangen commented:
done:
https://github.com/storybookjs/test-runner/pull/250
